Library/Homebrew/cli/args.rb

```diff
@@ -241,6 +240,51 @@ def spec(default = :stable)
           default
         end
       end
+
+      def resolve_keg(name)
+        require "keg"
+        require "formula"
+        require "missing_formula"
+
+        raise UsageError if name.empty?
+
+        rack = Formulary.to_rack(name.downcase)
+
+        dirs = rack.directory? ? rack.subdirs : []
+
+        raise NoSuchKegError, rack.basename if dirs.empty?
+
+        linked_keg_ref = HOMEBREW_LINKED_KEGS/rack.basename
+        opt_prefix = HOMEBREW_PREFIX/"opt/#{rack.basename}"
+
+        begin
+          if opt_prefix.symlink? && opt_prefix.directory?
+            Keg.new(opt_prefix.resolved_path)
+          elsif linked_keg_ref.symlink? && linked_keg_ref.directory?
+            Keg.new(linked_keg_ref.resolved_path)
+          elsif dirs.length == 1
+            Keg.new(dirs.first)
+          else
+            f = if name.include?("/") || File.exist?(name)
+              Formulary.factory(name)
+            else
+              Formulary.from_rack(rack)
+            end
+
+            unless (prefix = f.installed_prefix).directory?
+              raise MultipleVersionsInstalledError, rack.basename
+            end
+
+            Keg.new(prefix)
+          end
+        rescue FormulaUnavailableError
```

Comment:
Is this definitely the right error here?

Reply:
I'm not sure, and frankly I'm worried about changing it. I traced this line back to 2012, added in [this commit](https://github.com/Homebrew/brew/commit/9ab1df5b6875e9acc22f5bc4358373104af1e790)

To be honest, I'm unsure under what circumstances this code is executed. Ostensibly, it's run when there are multiple kegs installed to the same rack, which as far as I understand means two versions of the same formula are installed simultaneously. I wrote a script to force this to happen:

```
LR=$pwd
cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/

echo "Installing youtube-dl 2020.05.08"
git checkout 877666acfa7af926b0623dcdc7839e182c8e95d4 -- Formula/youtube-dl.rb
brew install youtube-dl
git restore --staged Formula/youtube-dl.rb
git restore Formula/youtube-dl.rb

echo "Moving youtube-dl to ~/Desktop temporarily"
mv /usr/local/Cellar/youtube-dl/2020.05.08 ~/Desktop

echo "Installing youtube-dl latest"
brew install youtube-dl

echo "Moving youtube-dl 2020.05.08 into Cellar"
mv ~/Desktop/2020.05.08 /usr/local/Cellar/youtube-dl/

cd $LR
```

And the error message I got when I ran `brew uninstall youtube-dl` originated from `cmd/uninstall.rb:92:96` on my branch, so not even in the `args.rb` file:
```
if rack.directory?
  versions = rack.subdirs.map(&:basename)
  puts "#{keg.name} #{versions.to_sentence} #{"is".pluralize(versions.count)} still installed."
  puts "Run `brew uninstall --force #{keg.name}` to remove all versions."
end
```

I think the chronology looked something like this:
1. Initially, if you had multiple kegs installed to the same rack, then it would error out and force you to manually remove the old kegs.
2. Later, [brew uninstall was changed](https://github.com/Homebrew/brew/commit/b42f76939cebb7267edaeb4fc2cb8a00af4b2789) so that it would remove the latest version and notify you of any older versions that were still installed. The error logic mentioned in the above comment was buried was buried in the process, never to be executed again. 

Do you know of a way to get the above rescue block to execute? Ah, If only we lived in a universe where the Halting Problem was decidable...

Reply:
> I'm not sure, and frankly I'm worried about changing it. I traced this line back to 2012, added in [this commit](https://github.com/Homebrew/brew/commit/9ab1df5b6875e9acc22f5bc4358373104af1e790)

Good enough for me 👍